# importing libraries and data

In [2]:
import pandas as pd
import numpy as np
import os
import scipy
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
path=r"C:\Users\Sam\Desktop\CF\Project 4 Instacart Basket\Instacart Basket Analysis\02 Data\Original Data"

In [3]:
df=pd.read_csv(os.path.join(path,"customers.csv"))

In [4]:
# checking columns
df.head(3)

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568

In [5]:
# changing First Name, Surnam, Gender, STATE, Age, n_dependants, fam_status
df.rename(columns={"First Name":"first_name"},inplace=True)
df.rename(columns={"Surnam":"last_name"},inplace=True)
df.rename(columns={"Gender":"gender"},inplace=True)
df.rename(columns={"STATE":"state"},inplace=True)
df.rename(columns={"Age":"age"},inplace=True)
df.rename(columns={"n_dependants":"number_of_dependants"},inplace=True)
df.rename(columns={"fam_status":"marital_status"},inplace=True)

In [6]:
#checking for execution
df.head(1)

user_id first_name last_name  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665

In [7]:
# getting datatypes
df.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
income                   int64
dtype: object

In [8]:
# checking the shape
df.shape

(206209, 10)

In [9]:
# descriptive stats
df.describe()

user_id            age  number_of_dependants         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [10]:
# checking if there are any missing values at all
df.isnull().values.any()

True

In [11]:
# since there are missing values, how many in total
df.isnull().sum().sum()

11259

In [12]:
# checking missing values by variable
df.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
marital_status              0
income                      0
dtype: int64

In [13]:
# checking for duplicates
df_dups=df[df.duplicated()]

In [14]:
df_dups.head()

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, marital_status, income]
Index: []

In [15]:
df_dups.shape

(0, 10)

In [16]:
# checling for mixed values
for col in df.columns.tolist():
  weird = (df[[col]].map(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)
else:
    print("No mixed values")
    

first_name
No mixed values


In [17]:
df["first_name"].dtype

dtype('O')

# Consistency check observations

## 1) I have decided to rename all variables to be consistent. I dont think it is a good idea to drop any of the variables, I might need the information in all of them. 2) All datatypes seem to be what they are supposed to be at least for now, if changes need to be made when merging will handle that then. 3) I found nothing abnormal in the descriptive stats. 4) There are missing values in the frame, all of them (11259) are in the "first_name" variable. To handle the missing values in this case, I think it would be best to replace the missing values with "noname" since it is the first name of the customer and we dont want to delete information that can be useful. 5) The dataframe has no duplicate values. 5) Checking for mixed data types showed that the coulmn "first_name" has mixed types, so I will change the datatype for the variable to a string.

In [18]:
# replacing missing values in "first_name"
df["first_name"].fillna("Noname",inplace=True)

In [19]:
# checking for execution
df.isnull().values.any()

False

In [20]:
# changing the datatype for "first_name"
df["first_name"]=df["first_name"].astype("str")

In [21]:
# checking for execution
for col in df.columns.tolist():
  weird = (df[[col]].map(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)
else:
    print("No mixed values")

No mixed values


In [22]:
# checking to see if string and object are different
df.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
income                   int64
dtype: object

### Quality and Consitency checks done

# Merging customer frame with the rest of cleaned instacart data

In [3]:
# importing cleaned data
path1=r"C:\Users\Sam\Desktop\CF\Project 4 Instacart Basket\Instacart Basket Analysis\02 Data\Prepared Data"
df_inst=pd.read_pickle(os.path.join(path1,"ords_prods_merge_revised_grouped.pkl"))

In [4]:
#df_inst.head()

In [5]:
#df.head()

### the two frames can be combined on a common key "user_id"

In [6]:
# checking user_id datatype in both frames
#df_inst["user_id"].dtype

In [7]:
#df["user_id"].dtype

#### both have the same datatype, we can proceed to merging.

In [8]:
#df_inst["_merge"].value_counts()

In [9]:
# dropping the variable
df_inst.drop(columns=["_merge"],inplace=True)

In [10]:
#df_inst.head()

## Tried merging the frames got memory error. Going to save the frames and start a new notebook with column dropped in df_inst

In [11]:
#df.to_pickle(os.path.join(path1,"customer_cleaned.pkl"))

In [12]:
df_inst.to_pickle(os.path.join(path1, "instacart_merge_dropped.pkl"))